ring.go

```diff
@@ -220,25 +222,94 @@ type ringShards struct {
 }
 
 func newRingShards(opt *RingOptions) *ringShards {
-	shards := make(map[string]*ringShard, len(opt.Addrs))
+	shards, list := createRingShards(opt, opt.Addrs)
+
+	c := &ringShards{
+		opt: opt,
+
+		shards: shards,
+		list:   list,
+	}
+	c.rebalance()
+
+	return c
+}
+
+func createRingShards(opt *RingOptions, addrs map[string]string) (map[string]*ringShard, []*ringShard) {
+	shards := make(map[string]*ringShard, len(addrs))
 	list := make([]*ringShard, 0, len(shards))
 
-	for name, addr := range opt.Addrs {
+	for name, addr := range addrs {
 		shard := newRingShard(opt, name, addr)
 		shards[name] = shard
 
 		list = append(list, shard)
 	}
+	return shards, list
+}
 
-	c := &ringShards{
-		opt: opt,
+func (c *ringShards) SetAddrs(addrs map[string]string) {
+	var currentRingshards map[string]*ringShard
+	toDelete := make(map[string]*ringShard)
+	newShards := make(map[string]*ringShard)
 
-		shards: shards,
-		list:   list,
+	c.mu.RLock()
+	if !c.closed {
```

Comment:
When closed I think it is better to simply return here

Reply:
agree